CONTRACTS: refactor DFCC code for loop contracts

remi-delmas-3000 opened the pull request:
This PR refactors the DFCC code to make it possible to handle loop assigns clauses in https://github.com/diffblue/cbmc/pull/7541.

We now have a standalone class `dfcc_contract_clauses_codegent` which provides methods to generate GOTO programs that initialize a dynamic frame instance from an assigns clause specification or havoc the targets of an assigns clause.

Before, these code generation methods were encapsulated inside the class `dfcc_contract_functionst` which exclusively handles function contracts. This makes them available for loop contracts as well.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Commit:
CONTRACTS: rearchitecture code for loop contracts

- Factor out some methods in dfcc_spec_functionst
- Move the code generation methods from dfcc_contract_functionst
  to dfcc_contract_clauses_codegent.
- Update Makefile with new class.
- Propagate interface changes to top level class

Commit:
CONTRACTS: new inlining and loop detection methods

Allows to to inlining and loop detection directly on a goto program
instead of a goto function

martin-cs reviewed (approved):
I have not reviewed the `contracts/dynamic-frames` part.  @remi-delmas-3000 I can if you want.  The rest looks good but I notice the overlap with #7550 so it would be good to have this rebased when it is merged.

Commit:
CONTRACTS: new context tracking flags in write set

Enable checking for the presence of dynamic
allocations or deallocations in loop bodies
and transitively in the functions called from
loop bodies.

- Adds `allow_allocate` and `allow_deallocate`
  flags in the write set struct
- Modify `add_allocated` to fail an assertion when
  called on a write set with `allow_allocate`
  set to false.
- Modify `check_deallocate` to return false when
  called on a write set with `allow_deallocate`
  set to false.
- Don't treat replacement mode specially anymore,
  always store freeable targets in both the object
  map (for lookups) and in the append list
  (for inclusion checks).

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc.cpp:
```diff
@@ -483,6 +490,8 @@ void dfcct::instrument_other_functions()
 
   goto_model.goto_functions.update();
 
+  // TODO specialise the library functions for the max size of
+  // loop and function contracts
```
Shall we create a GitHub issue about this and add the link here? It might be better to track this later.

remi-delmas-3000 replied:
this is still work in progress this will already be addressed in the final version

remi-delmas-3000 replied:
> DELETED

tautschnig replied:
> But then ... haven't you resolved the "TODO" and should remove this comment?

This has to be implemented in the forthcoming loop contracts PR, this TODO is just a marker that tells you where to do it

remi-delmas-3000 replied:
Sorry this one is a TODO for @qinheping , my comment above applied to two other TODOs I've turned into reqs.

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_clauses_codegen.cpp:
````diff
@@ -0,0 +1,293 @@
+  if(can_cast_expr<side_effect_expr_function_callt>(target))
+  {
+    // A function call target `foo(params)` becomes `CALL foo(params);`
+    // ```
````
can you remove this line?

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_clauses_codegen.cpp:
```diff
@@ -0,0 +1,293 @@
+{
+  const source_locationt &source_location = group.source_location();
+
+  // clean up side effects from the condition expression if needed
+  cleanert cleaner(goto_model.symbol_table, log.get_message_handler());
+  exprt condition(group.condition());
+  if(has_subexpr(condition, ID_side_effect))
+    cleaner.clean(condition, dest, language_mode);
```
nitpick: we might want to add this to a utility function.

remi-delmas-3000 replied:
The cleaner is only used in that single place in the file so maybe not that useful to factor it out ?

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -296,33 +361,28 @@ void dfcc_spec_functionst::to_spec_assigns_function(
 
-void dfcc_spec_functionst::to_spec_frees_function(
-  const irep_idt &function_id,
+// TODO add write_set_to_check an side effect checking
```
Shall we also create a GitHub issue for this one?

remi-delmas-3000 replied:
this is will be addressed before the final version of the PR

feliperodri reviewed (approved):
LGTM.

Commit:
CONTRACTS: new c++ methods to build calls to cprover_contracts library

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_clauses_codegen.cpp:
```diff
@@ -0,0 +1,288 @@
+{
+  for(const auto &expr : assigns_clause)
+  {
+    if(can_cast_expr<conditional_target_group_exprt>(expr))
+    {
+      encode_assignable_target_group(
+        language_mode, to_conditional_target_group_expr(expr), dest);
```
You might want to use `if(auto target_group = expr_try_dynamic_cast<conditional_target_group_exprt>(expr))` and then just use `*target_group` instead of `to_conditional_target_group_expr(expr)`.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -184,85 +236,78 @@ void dfcc_spec_functionst::generate_havoc_function(
+  std::set<irep_idt> function_pointer_contracts;
+  instrument.instrument_function(function_id, function_pointer_contracts);
+  INVARIANT(
+    function_pointer_contracts.size() == 0,
```
You might want to use `.empty()` instead of `.size() == 0`.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -184,85 +236,78 @@ void dfcc_spec_functionst::generate_havoc_function(
           "Function pointer call '" +
-            from_expr(ns, function_id, ins_it->call_function()) +
-            "' in function '" + id2string(function_id) + "' is not supported",
+            from_expr(ns, "", ins_it->call_function()) +
```
Without an identifier we cannot determine the mode, i.e., the language. It might then be better to fully fall back to `format_to_string`, which is language agnostic.

remi-delmas-3000 replied:
Added the language mode as extra parameter

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -184,85 +236,78 @@ void dfcc_spec_functionst::generate_havoc_function(
-            from_expr(ns, function_id, ins_it->call_function()) +
-            "' in function '" + id2string(function_id) + "' is not supported",
+            from_expr(ns, "", ins_it->call_function()) +
+            "' are supported in assigns clauses",
```
I am no longer able to parse this error message.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -296,33 +360,27 @@ void dfcc_spec_functionst::to_spec_assigns_function(
           "Function pointer call '" +
-            from_expr(ns, function_id, ins_it->call_function()) +
-            "' in function '" + id2string(function_id) + "' is not supported",
+            from_expr(ns, "", ins_it->call_function()) +
+            "' are not supported in frees clauses",
```
As above: consider using `format_to_string` instead, and the grammar seems broken.

remi-delmas-3000 replied:
now passing the language mode explicitly

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_wrapper_program.cpp:
```diff
@@ -30,6 +30,14 @@ Author: Remi Delmas, delmasrd@amazon.com
 #include "dfcc_lift_memory_predicates.h"
 #include "dfcc_utils.h"
 
+/// Lift a c++ bool to an exprt
+static exprt to_bool_expr(bool v)
+{
+  if(v)
+    return true_exprt();
+  return false_exprt();
+}
+
```
No need: use `make_boolean_expr` (from `expr_util.{h,cpp}`).

remi-delmas-3000 replied:
fixed

tautschnig reviewed (approved):
Approving with a couple of comments that really need to be addressed.

Commit:
address review comments
